### PR TITLE
Introduce `Logger`, a simple yet powerful logging system

### PR DESCRIPTION
**Pull request by Morriar**
# A simple logger for Nit

## Basic Usage

Create a new `Logger` with a severity level threshold set to `warn_level`:

~~~nit
var logger = new Logger(warn_level)
~~~

Messages with a severity equal or higher than `warn_level` will be displayed:

~~~nit
logger.error "Displays an error."
logger.warn "Displays a warning."
~~~

Messages with a lower severity are silenced:

~~~nit
logger.info "Displays nothing."
~~~

`FileLogger` can be used to output the messages into a file:

~~~nit
var log_file = "my.log"

logger = new FileLogger(warn_level, log_file, append = false)
logger.error("An error")
logger.info("Some info")
logger.close

assert log_file.to_path.read_all == "An error\n"
log_file.to_path.delete
~~~

## Severity levels

Each message is associated with a level that indicate its severity.
Only messages with a severity equal to or higher than the logger `level`
threshold will be displayed.

Severity levels from the most severe to the least severe:

* `unknown_level`: An unknown message that should always be outputted.
* `fatal_level`: An unhandleable error that results in a program crash.
* `error_level`: A handleable error condition.
* `warn_level`: A warning.
* `info_level`: Generic (useful) information about system operation.
* `debug_level`: Low-level information for developpers.

## Formatting messages

You can create custom formatters by implementing the `Formatter` interface.

~~~nit
class MyFormatter
	super Formatter

	redef fun format(level, message) do
		if level < warn_level then return message
		return "!!!{message}!!!"
	end
end
~~~

See `DefaultFormatter` for a more advanced implementation example.

Each Logger can be given a default formatter used to format the every messages
before outputting them:

~~~nit
var formatter = new MyFormatter
var stderr = new StringWriter
var logger = new Logger(warn_level, stderr, formatter)

logger.warn("This is a warning.")
assert stderr.to_s.trim.split("\n").last == "!!!This is a warning.!!!"
~~~

Optionally, a `Formatter` can be given to replace the `default_formatter`
used by default:

~~~nit
# Create a formatter with no default decorator
logger = new Logger(warn_level, stderr, null)

# Display a message without any formatter
logger.warn("This is a warning.")
assert stderr.to_s.trim.split("\n").last == "This is a warning."

# Display a message with a custom formatter
logger.warn("This is a warning.", formatter)
assert stderr.to_s.trim.split("\n").last == "!!!This is a warning.!!!"
~~~

### CODE REVIEW

**Comment by Morriar** [comment]
test this please

File: lib/logger/logger.nit [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+# ## Formatting messages
+#
+# You can create custom formatters by implementing the `Formatter` interface.
+#
+# ~~~
+# class MyFormatter
+#	super Formatter
+#
+#	redef fun format(level, message) do
+#		if level < warn_level then return message
```
Comment (privat) [inline lib/logger/logger.nit]: `return super` ?

File: lib/logger/logger.nit [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+	# var logger = new Logger(warn_level, stderr, null)
+	#
+	# # This message will be displayed:
+	# assert logger.warn("This is a warning.")
+	# assert stderr.to_s.trim.split("\n").last == "This is a warning."
+	#
+	# # This message will not:
+	# assert not logger.info("This is some info.")
+	# assert stderr.to_s.trim.split("\n").last == "This is a warning."
+	# ~~~
```
Comment (privat) [inline lib/logger/logger.nit]: This doc (and example) seems unrelated to `add_raw`

File: lib/logger/logger.nit [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+	#
+	# If `append` is `false`, the `file` will be overwritten.
+	var append: Bool = true is optional
+
+	init do
+		var old = null
+		if append then
+			old = file.to_path.read_all
+		end
+		out = new FileWriter.open(file)
```
Comment (privat) [inline lib/logger/logger.nit]: Oh, it seems we do not have an `append` mode on files in Nit.

File: lib/logger/logger.nit [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+
+# Format messages before outputing them
+#
+# A `Logger` can use a `Formatter` to format the messages before outputting them.
+#
+# See `DefaultFormatter`.
+interface Formatter
+
+	# Format `message` depending of its severity `level`
+	fun format(level: Int, message: Writable): Writable is abstract
```
Comment (privat) [inline lib/logger/logger.nit]: `do return message` seems a good default behavior

File: lib/logger/logger.nit [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+end
+
+redef class Sys
+
+	# Unknown severity level
+	#
+	# These messages are always displayed.
+	#
+	# See `Logger`.
+	var unknown_level = 5
```
Comment (privat) [inline lib/logger/logger.nit]: we need enums. (one day!)

File: lib/logger/logger.nit (hunk 4 above)
Comment (Morriar) [reply lib/logger/logger.nit]: Done.

File: lib/logger/logger.nit (hunk 1 above)
Comment (Morriar) [reply lib/logger/logger.nit]: Done.

File: lib/logger/logger.nit (hunk 3 above)
Comment (Morriar) [reply lib/logger/logger.nit]: I was surprised too.

Comment (Morriar) [reply lib/logger/logger.nit]: There is still an issue about the `stream` lib tho: https://github.com/nitlang/nit/issues/466

File: lib/logger/logger.nit (hunk 2 above)
Comment (Morriar) [reply lib/logger/logger.nit]: Good catch, changed the example.

**Comment by Morriar** [comment]
Reroll: I answered in each conversation.

**Comment by Morriar** [comment]
Tests are ok: https://gitlab.com/nit/nit-ci/pipelines/66653845
